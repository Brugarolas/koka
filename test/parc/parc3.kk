// Test value boxing
fun foo(x, y) {
  /* drop y */
  x
}

fun match1(xs:list<maybe<int>>, y:int) {
  match(xs) {
    Cons(Just(x)) | x == y -> /* dup x; drop xs; drop y */ x
    Cons(Nothing) -> /* drop xs; drop y */ 2
    _ -> /* drop xs */ y
  }
}
