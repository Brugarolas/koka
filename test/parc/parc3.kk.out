module parc/parc3
private import std/core/types = std/core/types = "";
private import std/core/hnd = std/core/hnd = "";
private import std/core = std/core = "";
public fun test : (xs : list<std/core/types/int>, y : std/core/types/int, z : std/core/types/int) -> std/core/types/int
 = fun(xs: list<std/core/types/int>, y: std/core/types/int, z: std/core/types/int){
 match (xs) {
 (std/core/Cons(((.skip std/core/types/.Box((x: std/core/types/int)) : std/core/types/.Box ) as .box-x96: std/core/types/.Box), (.pat0: list<
 std/core/types/int>)) : list<std/core/types/int> )
 | std/core/(==.1)((std/core/types/.dup(x)), (std/core/types/.dup(z))) -> val _ : ()
 = std/core/types/.drop(y);
 val _ : ()
 = std/core/types/.drop(z);
 val _ : ()
 = (match ((std/core/types/.is-unique(xs))) {
 (std/core/types/True() : std/core/types/bool )
 -> std/core/types/.drop(.pat0);
 std/core/types/.free(xs);
 _
 -> std/core/types/.dup(x);
 std/core/types/.dec-ref(xs);
 });
 x;
 (std/core/Cons(((.skip std/core/types/.Box((x0: std/core/types/int)) : std/core/types/.Box ) as .box-x97: std/core/types/.Box), (.pat2: list<
 std/core/types/int>)) : list<std/core/types/int> )
 | std/core/(==.1)((std/core/types/.dup(x0)), (std/core/types/.dup(y))) -> val _ : ()
 = std/core/types/.drop(xs);
 val _ : ()
 = std/core/types/.drop(z);
 y;
 (std/core/Cons(((.skip std/core/types/.Box((.pat4: std/core/types/int)) : std/core/types/.Box ) as .box-x98: std/core/types/.Box), (.pat5: list<
 std/core/types/int>)) : list<std/core/types/int> )
 | std/core/(==.1)((std/core/types/.dup(y)), (std/core/types/.dup(z))) -> val _ : ()
 = std/core/types/.drop(xs);
 val _ : ()
 = std/core/types/.drop(y);
 val _ : ()
 = std/core/types/.drop(z);
 2;
 _
 -> val _ : ()
 = std/core/types/.drop(xs);
 val _ : ()
 = std/core/types/.drop(z);
 y;
 }
 };